TeachingRecordSystem/src/TeachingRecordSystem.Api/V3/Core/Operations/SetQtls.cs

```diff
@@ -0,0 +1,127 @@
+using Microsoft.Xrm.Sdk.Query;
+using TeachingRecordSystem.Api.V3.Core.SharedModels;
+using TeachingRecordSystem.Api.Validation;
+using TeachingRecordSystem.Core.Dqt;
+using TeachingRecordSystem.Core.Dqt.Models;
+using TeachingRecordSystem.Core.Dqt.Queries;
+
+namespace TeachingRecordSystem.Api.V3.Core.Operations;
+
+public record SetQtlsCommand(string Trn, DateOnly? QtsDate);
+
+public sealed class SetQtlsResult
+{
+    private SetQtlsResult() { }
+
+    public bool Succeeded { get; private set; }
+
+    public QtlsInfo? QtlsInfo { get; private set; }
+
+    public static SetQtlsResult Success(QtlsInfo qtlsInfo) => new()
+    {
+        Succeeded = true,
+        QtlsInfo = qtlsInfo
+    };
+
+    public static SetQtlsResult Failed() => new()
+    {
+        Succeeded = false
+    };
+}
+
+public class SetQtlsHandler(ICrmQueryDispatcher _crmQueryDispatcher, IClock _clock)
+{
+    public async Task<SetQtlsResult> Handle(SetQtlsCommand command)
+    {
+        var contact = (await _crmQueryDispatcher.ExecuteQuery(
+            new GetActiveContactByTrnQuery(
+                command.Trn,
+                new ColumnSet(
+                    Contact.Fields.dfeta_TRN,
+                    Contact.Fields.dfeta_InductionStatus,
+                    Contact.Fields.dfeta_qtlsdate,
+                    Contact.Fields.dfeta_QTSDate,
+                    Contact.Fields.dfeta_ActiveSanctions
+                    )
+                )
+            ))!;
+
+        if (contact == null)
+        {
+            throw new ErrorException(ErrorRegistry.TeacherWithSpecifiedTrnNotFound());
+        }
+
+        var induction = await GetInductionWithAppropriateBody(contact.Id);
+        var (canSetQtlsDate, reviewTaskDescription) = CanSetQtlsDate(hasQTS: contact.dfeta_QTSDate.HasValue, overallInductionStatus: contact.dfeta_InductionStatus, inductionStatus: induction?.Induction.dfeta_InductionStatus, hasInductionWithAB: induction?.HasAppropriateBody ?? false, existingQtlsdate: contact.dfeta_qtlsdate, incomingQtlsDate: command.QtsDate);
+        if (!canSetQtlsDate)
+        {
+            await _crmQueryDispatcher.ExecuteQuery(
+                new CreateReviewTaskQuery()
+                {
+                    ContactId = contact.Id,
+                    Category = "Unable to set QTLSDate",
+                    Description = reviewTaskDescription!,
+                    Subject = "Notification for SET QTLS data collections team",
+                    ScheduledEnd = _clock.UtcNow
+                }
+            );
+
+            return SetQtlsResult.Failed();
+        }
+
+        if (contact.dfeta_ActiveSanctions == true)
+        {
+            await _crmQueryDispatcher.ExecuteQuery(
```

Comment:
This should really be within the `SetQtlsDateHandler` so that it can be made atomic with the date change.